File: app/interceptor/matcher.py
"""Rule matcher (arch §3.3/§4.1, AC-9, AC-33b).

Selects the first enabled rule that matches a request, ordered by ``priority``
(lower first) then ``id`` — deterministic (§5.3 ordering). Matching covers:

  * **method**     — ``ANY`` or an exact (case-insensitive) HTTP verb.
  * **path**       — exact, ``:param`` segments (captured for templating), and a
                     trailing ``/*`` wildcard. Compiled once to a regex.
  * **headers**    — required header name→value (case-insensitive name).
  * **query**      — required query key→value.
  * **body**       — jsonpath-lite ``eq/neq/contains/exists`` conditions.
  * **state**      — ``state_requirements`` (``eq/neq/exists/absent``) — gates the
                     rule on per-endpoint Redis state. **Fail-CLOSED** (AC-9/AC-49):
                     if state is unavailable (Redis down) a state-gated rule is
                     **skipped**, never silently matched.

Path captures (``:name``) are returned so the templating engine can resolve
``{{request.path.<name>}}`` (§5.7).
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import Dict, List, Optional, Pattern, Tuple

from app.models import MatchCriteria
from app.utils.helpers import jsonpath_lite


def compile_path(path: str) -> Pattern[str]:
    """Compile a match path into an anchored regex.

    ``/users/:id`` -> capture group ``id``; trailing ``/*`` -> match any suffix;
    everything else is matched literally. The match path is normalized to start
    with ``/``.
    """
    if not path:
        path = "/*"
    if not path.startswith("/"):
        path = "/" + path

    # Pure catch-all: match any path. (Handle before the trailing-/* strip below.)
    if path in ("/*", "*", "/**"):
        return re.compile(r"^/.*$")

    wildcard = False
    if path.endswith("/*"):
        wildcard = True
        path = path[:-2]  # strip the trailing "/*"

    segments = [s for s in path.split("/") if s != ""]
    parts: List[str] = []
    for seg in segments:
        if seg.startswith(":") and len(seg) > 1:
            name = seg[1:]
            # A capture is a single non-slash segment.
            parts.append(rf"/(?P<{re.escape(name)}>[^/]+)")
        else:
            parts.append("/" + re.escape(seg))

    body = "".join(parts) if parts else "/"
    if wildcard:
        # Prefix followed by anything (a further "/..." suffix) or nothing.
        pattern = rf"^{body}(?:/.*)?$" if body != "/" else r"^/.*$"
    else:
        pattern = rf"^{body}/?$" if body != "/" else r"^/$"
    return re.compile(pattern)


@dataclass
class CompiledRule:
    """A rule compiled for the fast path (precompiled path regex + parsed criteria)."""

    id: int
    priority: int
    enabled: bool
    method: str                       # upper-cased; "ANY" matches all
    path_pattern: Pattern[str]
    raw_path: str
    headers: Dict[str, str]           # lower-cased names
    query: Dict[str, str]
    body_conditions: List[dict]       # [{path, op, value}]
    state_requirements: List[dict]    # [{key, op, value}]
    response: dict                    # ResponseSpec dict
    state_writes: List[dict]          # [{key, value}]
    latency_ms: Optional[int]
    rate_limit_per_min: Optional[int]
    webhook_action: Optional[dict]

    @property
    def gates_on_state(self) -> bool:
        return bool(self.state_requirements)


def compile_rule(rule: dict) -> CompiledRule:
    """Build a :class:`CompiledRule` from a DB-shaped rule dict (already JSON-parsed
    ``match``/``response``/``state_writes``)."""
    match = rule.get("match") or {}
    if isinstance(match, MatchCriteria):
        match = match.model_dump()
    headers = {k.lower(): v for k, v in (match.get("headers") or {}).items()}
    return CompiledRule(
        id=int(rule["id"]),
        priority=int(rule.get("priority", 100)),
        enabled=bool(rule.get("enabled", True)),
        method=str(match.get("method", "ANY")).upper(),
        path_pattern=compile_path(match.get("path", "/*")),
        raw_path=match.get("path", "/*"),
        headers=headers,
        query=dict(match.get("query") or {}),
        body_conditions=list(match.get("body_conditions") or []),
        state_requirements=list(match.get("state_requirements") or []),
        response=dict(rule.get("response") or {}),
        state_writes=list(rule.get("state_writes") or []),
        latency_ms=rule.get("latency_ms"),
        rate_limit_per_min=rule.get("rate_limit_per_min"),
        webhook_action=rule.get("webhook_action"),
    )


# --- individual predicate helpers ---------------------------------------------
def _method_ok(rule: CompiledRule, method: str) -> bool:
    return rule.method == "ANY" or rule.method == method.upper()


def _path_ok(rule: CompiledRule, path: str) -> Optional[Dict[str, str]]:
    m = rule.path_pattern.match(path)
    if m is None:
        return None
    return m.groupdict()


def _headers_ok(rule: CompiledRule, headers: Dict[str, str]) -> bool:
    for name, want in rule.headers.items():
        if headers.get(name.lower()) != want:
            return False
    return True


def _query_ok(rule: CompiledRule, query: Dict[str, str]) -> bool:
    for key, want in rule.query.items():
        if query.get(key) != want:
            return False
    return True


def _body_ok(rule: CompiledRule, body: str) -> bool:
    for cond in rule.body_conditions:
        path = cond.get("path", "")
        op = cond.get("op", "eq")
        want = cond.get("value")
        got = jsonpath_lite(body, path)
        if op == "exists":
            if got is None:
                return False
        elif op == "eq":
            if got != (want or ""):
                return False
        elif op == "neq":
            if got == (want or ""):
                return False
        elif op == "contains":
            if got is None or (want or "") not in got:
                return False
        else:
            return False
    return True


def _state_ok(rule: CompiledRule, state: Dict[str, str]) -> bool:
    """Evaluate state requirements against the (possibly empty) state snapshot.

    When ``state`` is empty because Redis was unavailable, an ``eq``/``contains``
    requirement simply does not hold → the rule is skipped (fail-CLOSED, AC-9/49).
    """
    for req in rule.state_requirements:
        key = req.get("key", "")
        op = req.get("op", "eq")
        want = req.get("value")
        present = key in state
        got = state.get(key)
        if op == "exists":
            if not present:
                return False
        elif op == "absent":
            if present:
                return False
        elif op == "eq":
            if not present or got != (want or ""):
                return False
        elif op == "neq":
            # neq holds if absent or differs.
            if present and got == (want or ""):
                return False
        else:
            return False
    return True


@dataclass
class MatchResult:
    rule: CompiledRule
    path_params: Dict[str, str] = field(default_factory=dict)


def select(
    rules: List[CompiledRule],
    method: str,
    path: str,
    headers: Dict[str, str],
    query: Dict[str, str],
    body: str,
    state: Dict[str, str],
) -> Optional[MatchResult]:
    """Return the first enabled, fully-matching rule (rules are pre-sorted by
    ``priority`` then ``id``), with its captured path params. ``None`` if none match.

    ``state`` is the (lazily-read) per-endpoint snapshot; pass ``{}`` when no rule
    gates on state or when Redis is down (state-gated rules then fail closed).
    """
    for rule in rules:
        if not rule.enabled:
            continue
        if not _method_ok(rule, method):
            continue
        params = _path_ok(rule, path)
        if params is None:
            continue
        if not _headers_ok(rule, headers):
            continue
        if not _query_ok(rule, query):
            continue
        if not _body_ok(rule, body):
            continue
        if rule.gates_on_state and not _state_ok(rule, state):
            continue
        return MatchResult(rule=rule, path_params=params)
    return None

File: backend/src/interceptor/matcher.rs
//! Rule matcher — PORT of `app/interceptor/matcher.py` (AC-11, AC-13, §5.3/§5.5).
//!
//! Selects the first **enabled** rule that fully matches a request, with rules
//! pre-sorted by `priority` (lower first) then `id` — deterministic ordering.
//! Matching covers method / path (exact · `:param` · trailing `/*`) / required
//! headers (case-insensitive name) / required query / `body_conditions`
//! (jsonpath-lite `eq|neq|contains|exists`) / `state_requirements`
//! (`eq|neq|exists|absent`).
//!
//! Path patterns are compiled into a segment matcher (no regex dependency) so a
//! crafted pattern cannot cause regex blowup. `:name` segments are captured for
//! `{{request.path.<name>}}` templating (§5.7).
//!
//! State-gated rules **fail closed** (AC-11/AC-57): when `state` is empty because
//! the store was unreadable, an `eq`/`contains`/`exists` requirement does not
//! hold → the rule is skipped, never silently matched.

use std::collections::BTreeMap;

use crate::helpers::jsonpath_lite;
use crate::models::{
    BodyCondition, MatchCriteria, ResponseSpec, StateRequirement, StateWrite, WebhookAction,
};

/// A compiled path: a sequence of segment matchers plus a trailing-wildcard flag.
#[derive(Debug, Clone)]
pub struct CompiledPath {
    segments: Vec<Segment>,
    /// Trailing `/*` — match this prefix followed by anything (or nothing).
    wildcard: bool,
    /// Pure catch-all (`/*`, `*`, `/**`) — matches any path.
    catch_all: bool,
}

#[derive(Debug, Clone)]
enum Segment {
    Literal(String),
    /// `:name` — captures one non-empty, non-slash segment under `name`.
    Param(String),
}

/// Compile a match path (`/users/:id`, `/v1/*`, `/exact`) into a `CompiledPath`.
/// Mirrors `matcher.py::compile_path` semantics.
pub fn compile_path(path: &str) -> CompiledPath {
    let mut p = if path.is_empty() {
        "/*".to_string()
    } else {
        path.to_string()
    };
    if !p.starts_with('/') {
        p = format!("/{p}");
    }
    if p == "/*" || p == "*" || p == "/**" {
        return CompiledPath {
            segments: Vec::new(),
            wildcard: false,
            catch_all: true,
        };
    }
    let mut wildcard = false;
    if let Some(stripped) = p.strip_suffix("/*") {
        wildcard = true;
        p = stripped.to_string();
    }
    let segments: Vec<Segment> = p
        .split('/')
        .filter(|s| !s.is_empty())
        .map(|seg| {
            if seg.len() > 1 && seg.starts_with(':') {
                Segment::Param(seg[1..].to_string())
            } else {
                Segment::Literal(seg.to_string())
            }
        })
        .collect();
    CompiledPath {
        segments,
        wildcard,
        catch_all: false,
    }
}

impl CompiledPath {
    /// Match `path` against this compiled pattern; return captured params
    /// (`:name` -> value) on a match, `None` otherwise.
    pub fn match_path(&self, path: &str) -> Option<BTreeMap<String, String>> {
        if self.catch_all {
            return Some(BTreeMap::new());
        }
        let parts: Vec<&str> = path.split('/').filter(|s| !s.is_empty()).collect();
        let n = self.segments.len();
        if self.wildcard {
            // Prefix must match; anything (incl. nothing) may follow.
            if parts.len() < n {
                return None;
            }
        } else if parts.len() != n {
            return None;
        }
        let mut params = BTreeMap::new();
        for (i, seg) in self.segments.iter().enumerate() {
            match seg {
                Segment::Literal(lit) => {
                    if parts[i] != lit {
                        return None;
                    }
                }
                Segment::Param(name) => {
                    params.insert(name.clone(), parts[i].to_string());
                }
            }
        }
        Some(params)
    }
}

/// A rule compiled for the fast path (parsed criteria + precompiled path).
#[derive(Debug, Clone)]
pub struct CompiledRule {
    pub id: i64,
    pub priority: i64,
    pub enabled: bool,
    pub method: String, // upper-cased; "ANY" matches all
    pub path: CompiledPath,
    pub headers: BTreeMap<String, String>, // lower-cased names
    pub query: BTreeMap<String, String>,
    pub body_conditions: Vec<BodyCondition>,
    pub state_requirements: Vec<StateRequirement>,
    pub response: ResponseSpec,
    pub state_writes: Vec<StateWrite>,
    pub latency_ms: Option<i64>,
    pub rate_limit_per_min: Option<i64>,
    pub chaos_mode: Option<String>,
    pub webhook_action: Option<WebhookAction>,
}

impl CompiledRule {
    pub fn gates_on_state(&self) -> bool {
        !self.state_requirements.is_empty()
    }
}

/// Build a `CompiledRule` from already-parsed DB fields.
#[allow(clippy::too_many_arguments)]
pub fn compile_rule(
    id: i64,
    priority: i64,
    enabled: bool,
    match_: &MatchCriteria,
    response: ResponseSpec,
    state_writes: Vec<StateWrite>,
    latency_ms: Option<i64>,
    rate_limit_per_min: Option<i64>,
    chaos_mode: Option<String>,
    webhook_action: Option<WebhookAction>,
) -> CompiledRule {
    let headers = match_
        .headers
        .iter()
        .map(|(k, v)| (k.to_ascii_lowercase(), v.clone()))
        .collect();
    CompiledRule {
        id,
        priority,
        enabled,
        method: match_.method.to_ascii_uppercase(),
        path: compile_path(&match_.path),
        headers,
        query: match_.query.clone(),
        body_conditions: match_.body_conditions.clone(),
        state_requirements: match_.state_requirements.clone(),
        response,
        state_writes,
        latency_ms,
        rate_limit_per_min,
        chaos_mode,
        webhook_action,
    }
}

fn method_ok(rule: &CompiledRule, method: &str) -> bool {
    rule.method == "ANY" || rule.method == method.to_ascii_uppercase()
}

fn headers_ok(rule: &CompiledRule, headers: &BTreeMap<String, String>) -> bool {
    rule.headers
        .iter()
        .all(|(name, want)| headers.get(name) == Some(want))
}

fn query_ok(rule: &CompiledRule, query: &BTreeMap<String, String>) -> bool {
    rule.query
        .iter()
        .all(|(key, want)| query.get(key) == Some(want))
}

fn body_ok(rule: &CompiledRule, body: &str) -> bool {
    for cond in &rule.body_conditions {
        let got = jsonpath_lite(body, &cond.path);
        let want = cond.value.clone().unwrap_or_default();
        let ok = match cond.op.as_str() {
            "exists" => got.is_some(),
            "eq" => got.as_deref() == Some(want.as_str()),
            "neq" => got.as_deref() != Some(want.as_str()),
            "contains" => got.as_deref().map(|g| g.contains(&want)).unwrap_or(false),
            _ => false,
        };
        if !ok {
            return false;
        }
    }
    true
}

fn state_ok(rule: &CompiledRule, state: &BTreeMap<String, String>) -> bool {
    for req in &rule.state_requirements {
        let present = state.contains_key(&req.key);
        let got = state.get(&req.key);
        let want = req.value.clone().unwrap_or_default();
        let ok = match req.op.as_str() {
            "exists" => present,
            "absent" => !present,
            "eq" => present && got == Some(&want),
            // neq holds if absent OR differs.
            "neq" => !present || got != Some(&want),
            _ => false,
        };
        if !ok {
            return false;
        }
    }
    true
}

/// A matched rule plus its captured path params.
pub struct MatchResult<'a> {
    pub rule: &'a CompiledRule,
    pub path_params: BTreeMap<String, String>,
}

/// Return the first enabled, fully-matching rule (rules are pre-sorted by
/// `priority` then `id`) with its captured path params. `None` if none match.
pub fn select<'a>(
    rules: &'a [CompiledRule],
    method: &str,
    path: &str,
    headers: &BTreeMap<String, String>,
    query: &BTreeMap<String, String>,
    body: &str,
    state: &BTreeMap<String, String>,
) -> Option<MatchResult<'a>> {
    for rule in rules {
        if !rule.enabled {
            continue;
        }
        if !method_ok(rule, method) {
            continue;
        }
        let params = match rule.path.match_path(path) {
            Some(p) => p,
            None => continue,
        };
        if !headers_ok(rule, headers) {
            continue;
        }
        if !query_ok(rule, query) {
            continue;
        }
        if !body_ok(rule, body) {
            continue;
        }
        if rule.gates_on_state() && !state_ok(rule, state) {
            continue;
        }
        return Some(MatchResult {
            rule,
            path_params: params,
        });
    }
    None
}

#[cfg(test)]
mod tests {
    use super::*;

    fn rule(id: i64, priority: i64, method: &str, path: &str) -> CompiledRule {
        let mc = MatchCriteria {
            method: method.into(),
            path: path.into(),
            ..Default::default()
        };
        compile_rule(
            id,
            priority,
            true,
            &mc,
            ResponseSpec::default(),
            vec![],
            None,
            None,
            None,
            None,
        )
    }

    fn empty() -> BTreeMap<String, String> {
        BTreeMap::new()
    }

    #[test]
    fn path_exact_param_and_wildcard() {
        assert!(compile_path("/users").match_path("/users").is_some());
        assert!(compile_path("/users").match_path("/users/1").is_none());
        let p = compile_path("/users/:id");
        let caps = p.match_path("/users/42").unwrap();
        assert_eq!(caps.get("id").map(String::as_str), Some("42"));
        assert!(p.match_path("/users").is_none());
        let w = compile_path("/v1/*");
        assert!(w.match_path("/v1").is_some());
        assert!(w.match_path("/v1/a/b/c").is_some());
        assert!(w.match_path("/v2/a").is_none());
        assert!(compile_path("/*").match_path("/anything/here").is_some());
    }

    #[test]
    fn order_is_priority_then_id() {
        let rules = vec![
            rule(2, 50, "ANY", "/*"),
            rule(1, 10, "ANY", "/*"),
            rule(3, 10, "ANY", "/*"),
        ];
        // caller pre-sorts; emulate priority,id order.
        let mut sorted = rules;
        sorted.sort_by_key(|r| (r.priority, r.id));
        let m = select(&sorted, "GET", "/x", &empty(), &empty(), "", &empty()).unwrap();
        assert_eq!(m.rule.id, 1); // priority 10, lowest id
    }

    #[test]
    fn disabled_and_method_filter() {
        let mut r = rule(1, 10, "POST", "/*");
        r.enabled = false;
        assert!(select(&[r], "POST", "/x", &empty(), &empty(), "", &empty()).is_none());
        let r2 = rule(1, 10, "POST", "/*");
        assert!(select(
            std::slice::from_ref(&r2),
            "GET",
            "/x",
            &empty(),
            &empty(),
            "",
            &empty()
        )
        .is_none());
        assert!(select(&[r2], "post", "/x", &empty(), &empty(), "", &empty()).is_some());
    }

    #[test]
    fn body_and_state_conditions() {
        let mut r = rule(1, 10, "ANY", "/*");
        r.body_conditions = vec![BodyCondition {
            path: "kind".into(),
            op: "eq".into(),
            value: Some("vip".into()),
        }];
        assert!(select(
            &[r.clone()],
            "POST",
            "/x",
            &empty(),
            &empty(),
            r#"{"kind":"vip"}"#,
            &empty()
        )
        .is_some());
        assert!(select(
            &[r],
            "POST",
            "/x",
            &empty(),
            &empty(),
            r#"{"kind":"std"}"#,
            &empty()
        )
        .is_none());

        let mut sr = rule(2, 10, "ANY", "/*");
        sr.state_requirements = vec![StateRequirement {
            key: "logged_in".into(),
            op: "eq".into(),
            value: Some("1".into()),
        }];
        // fail-closed: empty state -> skip.
        assert!(select(&[sr.clone()], "GET", "/x", &empty(), &empty(), "", &empty()).is_none());
        let mut state = BTreeMap::new();
        state.insert("logged_in".to_string(), "1".to_string());
        assert!(select(&[sr], "GET", "/x", &empty(), &empty(), "", &state).is_some());
    }
}
